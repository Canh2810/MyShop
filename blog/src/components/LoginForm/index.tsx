'use client'

// Libs
import { memo, useCallback, useState } from 'react'
import { useForm, Controller, SubmitHandler } from 'react-hook-form'
import { useRouter } from 'next/navigation'

// Components
import { Button, TextField, Typography } from '../index'

// Types
import {
  ButtonVariants,
  ILoginForm,
  LoginFormField,
  CommonStatus,
  TextFieldTypes,
  TypoVariants,
} from '@/types'

// Constants
import { ERROR_MESSAGES, ROUTES } from '@/constants'

// Icons
import { Email, Lock } from '@/assets'

// Utils
import { checkEmail } from '@/utils'

// Hooks
import { loginService } from '@/services'
import { useAuthStore } from '@/stores'

const LoginForm = () => {
  const [errorMessage, setErrorMessage] = useState<string | null>(null)
  const loginFormField: LoginFormField[] = ['email', 'password']
  const { control, clearErrors, handleSubmit } = useForm<ILoginForm>({
    defaultValues: {
      email: '',
      password: '',
    },
  })
  const router = useRouter()
  const login = useAuthStore((state) => state.login)

  /**
   * Handle login
   * @param data email and password value
   */
  const handleSubmitLoginForm: SubmitHandler<ILoginForm> = useCallback(
    (data: ILoginForm) => {
      const response = loginService(data)
      if (response.status === CommonStatus.Success) {
        response.data && login(response.data)
        router.push(ROUTES.HOME)
      } else {
        setErrorMessage(response.message || '')
      }
    },
    [router],
  )

  return (
    <form
      onSubmit={handleSubmit(handleSubmitLoginForm)}
      className="relative w-[500px] flex flex-col gap-[28px] bg-white p-10 rounded-xl shadow-md"
    >
      <Typography
        component="h1"
        variant={TypoVariants.HeadingMedium}
        className="text-center"
      >
        LOGIN
      </Typography>
      {loginFormField.map((item) => {
        const title = item === 'email' ? 'Email' : 'Password'
        return (
          <Controller
            key={item}
            control={control}
            name={item}
            rules={{
              required: ERROR_MESSAGES.FIELD_REQUIRED(title),
              ...(item === 'email' && {
                validate: (value: string) => checkEmail(value),
              }),
            }}
            render={({ field: { value, onChange }, fieldState: { error } }) => (
              <TextField
                icon={item === 'email' ? <Email /> : <Lock />}
                id={item}
                type={
                  item === 'email'
                    ? TextFieldTypes.Text
                    : TextFieldTypes.Password
                }
                name={item}
                placeholder={title}
                value={value}
                errorMessage={error?.message}
                onChange={(data) => {
                  clearErrors(item)
                  onChange(data)
                }}
              />
            )}
          />
        )
      })}
      <Button variant={ButtonVariants.Container} type="submit">
        Login
      </Button>
      {errorMessage && (
        <p className="text-center absolute bottom-2 left-0 right-0  text-danger">
          {errorMessage}
        </p>
      )}
    </form>
  )
}

export default memo(LoginForm)
