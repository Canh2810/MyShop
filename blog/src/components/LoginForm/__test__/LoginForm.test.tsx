import { render, waitFor, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import LoginForm from '..'

const useRouterMock = {
  push: jest.fn(),
}
const loginMock = jest.fn()

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useRouter: () => useRouterMock,
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => 'mocked-callback-url'),
  })),
}))

// Mocking services
jest.mock('@/services', () => ({
  ...jest.requireActual('@/services'),
  login: () => loginMock,
}))

describe('LoginForm Component', () => {
  const setup = () => {
    const { getByPlaceholderText, getByText } = render(<LoginForm />)

    const emailInput = getByPlaceholderText('Email')
    const passwordInput = getByPlaceholderText('Password')
    const submitButton = getByText('Login')

    return {
      emailInput,
      passwordInput,
      submitButton,
    }
  }

  global.fetch = jest.fn()
  it('should render correctly', () => {
    const { container } = render(<LoginForm />)
    expect(container).toMatchSnapshot()
  })

  it('handles form submission with valid data', async () => {
    const { emailInput, passwordInput } = setup()

    await userEvent.type(emailInput, 'test@example.com')
    await userEvent.type(passwordInput, 'password')
    // await userEvent.click(submitButton)

    // waitFor(() => {
    //   expect(useRouterMock.push).toHaveBeenCalledWith(ROUTES.HOME)
    // })
  })

  it('should display an error message for an invalid email', async () => {
    const { emailInput, submitButton } = setup()

    await userEvent.type(emailInput, 'invalidemail')
    await userEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Invalid format of email, please check.'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing email', async () => {
    const { passwordInput, submitButton } = setup()

    await userEvent.type(passwordInput, 'password@123')
    await userEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Email is required'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing password', async () => {
    const { emailInput, submitButton } = setup()

    await userEvent.type(emailInput, 'test@example.com')
    await userEvent.click(submitButton)

    const errorMessage = await waitFor(() =>
      screen.getByText('Password is required'),
    )

    expect(errorMessage).toBeInTheDocument()
  })

  it('should display an error message for a missing email and password', async () => {
    const { submitButton } = setup()
    await userEvent.click(submitButton)

    const emailErrorMessage = await waitFor(() =>
      screen.getByText('Email is required'),
    )

    const passwordErrorMessage = await waitFor(() =>
      screen.getByText('Password is required'),
    )

    expect(emailErrorMessage).toBeInTheDocument()
    expect(passwordErrorMessage).toBeInTheDocument()
  })
})
