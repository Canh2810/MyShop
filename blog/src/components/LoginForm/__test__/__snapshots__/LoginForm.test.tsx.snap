// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component should render correctly 1`] = `
<div>
  <form
    class="relative w-[500px] flex flex-col gap-[28px] bg-white p-10 rounded-xl shadow-md"
  >
    <h1
      class="font-primary text-lg font-bold leading-6 text-dark-100 dark:text-light-400 text-center"
    >
      LOGIN
    </h1>
    <div
      class="relative flex items-center px-4  py-3 rounded-md text-sm border border-solid border-light-700 text-dark-50 bg-white dark:bg-dark-200  "
    >
      <input
        class="w-full bg-transparent text-light-500 focus:outline-0"
        id="email"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <svg
        fill="none"
        height="20"
        width="20"
      >
        <path
          clip-rule="evenodd"
          d="M3.438 4.375a.937.937 0 0 0-.938.938v9.375c0 .517.42.937.938.937h13.124c.518 0 .938-.42.938-.938V5.313a.937.937 0 0 0-.938-.938H3.438Zm-2.188.938c0-1.209.98-2.188 2.188-2.188h13.124c1.209 0 2.188.98 2.188 2.188v9.375c0 1.208-.98 2.187-2.188 2.187H3.438a2.188 2.188 0 0 1-2.187-2.188V5.313Z"
          fill="#696A75"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M3.882 5.866a.625.625 0 0 1 .877-.11L10 9.834l5.241-4.076a.625.625 0 1 1 .768.986l-5.625 4.375a.625.625 0 0 1-.768 0L3.991 6.743a.625.625 0 0 1-.11-.877Z"
          fill="#696A75"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="relative flex items-center px-4  py-3 rounded-md text-sm border border-solid border-light-700 text-dark-50 bg-white dark:bg-dark-200  "
    >
      <input
        class="w-full bg-transparent text-light-500 focus:outline-0"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <svg
        class="ionicon"
        height="20px"
        viewBox="0 0 512 512"
        width="20px"
      >
        <path
          d="M336 208v-95a80 80 0 0 0-160 0v95"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
        />
        <rect
          fill="none"
          height="272"
          rx="48"
          ry="48"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
          width="320"
          x="96"
          y="208"
        />
      </svg>
    </div>
    <button
      class=" rounded-md hover:opacity-70 disabled:opacity-40 px-5 py-3 bg-primary-500 text-white "
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
